service/history/ndc/workflow_resetter.go

```diff
@@ -609,22 +612,30 @@ func (r *workflowResetterImpl) reapplyContinueAsNewWorkflowEvents(
 	}
 
 	getNextEventIDBranchToken := func(runID string) (nextEventID int64, branchToken []byte, retError error) {
-		context, release, err := r.workflowCache.GetOrCreateWorkflowExecution(
-			ctx,
-			r.shardContext,
-			namespaceID,
-			&commonpb.WorkflowExecution{
-				WorkflowId: workflowID,
-				RunId:      runID,
-			},
-			locks.PriorityHigh,
-		)
-		if err != nil {
-			return 0, nil, err
+		var wfCtx workflow.Context
+		var err error
+
+		if runID == currentWorkflow.GetMutableState().GetWorkflowKey().RunID {
```

Comment:
for somebody who may read this code later it will be very hard to understand why, even with PR comments.
This function is called only from workflowResetterImpl.ResetWorkflow. 
I would suggest to move this code to an upper level.
In fact I would suggest to figure out who/why/where is locking "current workflow" and move this code to the same level. This (something somewhere is locking before calling ResetWorkflow) is wrong usage of lock. 
It may be hard/impossible to do given the complexity of the codebase, but worth a try.

Reply:
I totally agree with your comment that the locking logic is confusing and definitely does not follow best practices about how to manage locking and function calls. 

I think it's possible that part of the reason for splitting `reapplyContinueAsNewWorkflowEvents` into its own function is for unit testing, but as we see from Yichao's comment on this PR, the unit testing here is also flawed.

Moving this logic up would require refactoring the unit tests that cover this code, and it feels risky to do that correctly. I generally agree with your idea though, so idk